File: packages/orch-protocol/src/tools.ts
// ---- Tool & ToolSet protocol types ----
// Host-visible tool surface types.

import type { ToolCall } from "./messages.js";

// ---- Tool capabilities ----

/** Capability tags that describe what a tool can do. */
export type ToolCapability =
  | "workspace_read"
  | "workspace_write"
  | "process"
  | "network"
  | "view_image"
  | "update_plan"
  | "user_input"
  | "delegation";

/** When a tool is visible to the provider and/or searchable. */
export type ToolExposure = "direct" | "deferred" | "hidden";

/** Approval requirement for a tool. */
export type ToolApprovalRequirement = "never" | "on_request" | "always";

/** Executor ref: where the tool implementation lives. */
export interface ToolExecutorRef {
  kind: "native" | "host" | "remote" | "sandbox" | "mcp" | "orchestrator";
  target: string;
  extra?: Record<string, unknown>;
}

/** Metadata for tool documentation and policy. */
export interface ToolMetadata {
  title?: string;
  readOnly?: boolean;
  destructive?: boolean;
  mutatesWorkspace?: boolean;
  producesArtifact?: boolean;
}

/** A single tool definition — the canonical per-tool shape. */
export interface ToolDef {
  name: string;
  description: string;
  inputSchema: unknown;
  executor: ToolExecutorRef;
  executionMode?: "sequential" | "parallel";
  exposure?: ToolExposure;
  capabilities?: ToolCapability[];
  approval?: ToolApprovalRequirement;
  metadata?: ToolMetadata;
}

// ---- ToolSet types ----

/** ToolSet metadata. */
export interface ToolSetMetadata {
  source?: "builtin" | "host" | "mcp" | "plugin" | "dynamic" | "agent";
  tags?: string[];
}

/** A named, grouped capability surface. Tools are references, not inline definitions. */
export interface ToolSet {
  id: string;
  name: string;
  description?: string;
  tools: ToolSetToolRef[];
  /** Per-tool-set policy defaults (orchestrator level). */
  policy?: ToolSetPolicy;
  metadata?: ToolSetMetadata;
}

export type ToolSetEntry = ToolDef | ToolSetToolRef;

export type ToolSetToolRef = ProviderToolRef | ProviderNamespaceRef | OrchestratorControlRef;

export interface ProviderToolRef {
  kind: "provider_tool";
  providerId: string;
  toolName: string;
  alias?: string;
  policy?: Partial<ToolPolicy>;
}

export interface ProviderNamespaceRef {
  kind: "provider_namespace";
  providerId: string;
  namespace: string;
  alias?: string;
  policy?: Partial<ToolPolicy>;
}

export interface OrchestratorControlRef {
  kind: "orchestrator_control";
  action: string;
  alias?: string;
  policy?: Partial<ToolPolicy>;
}

// ---- Policy types ----

export type ToolSensitivity = "safe" | "sensitive" | "dangerous" | "dynamic";
export type ToolApprovalPolicy = "never" | "on_sensitive" | "always";
export type ToolExecutionMode = "parallel" | "sequential";
export type ToolFailureMode = "return_error" | "fail_task";

export interface ToolPolicy {
  sensitivity?: ToolSensitivity;
  approval?: ToolApprovalPolicy;
  timeoutMs?: number;
  executionMode?: ToolExecutionMode;
  failureMode?: ToolFailureMode;
}

export interface ToolSetPolicy {
  defaults?: Partial<ToolPolicy>;
  allowParallel?: boolean;
  maxConcurrentCalls?: number;
}

// ---- ToolProvider interface ----

/** Source classification for a tool provider. */
export type ToolProviderSource = "orch" | "host" | "workspace" | "mcp" | "plugin";

/** Context passed to provider.discover() to scope tool discovery. */
export interface ToolDiscoveryContext {
  agentId: string;
  taskId?: string;
  toolSetIds: string[];
  activeToolNames?: string[];
}

/** Context for tool execution. */
export interface ToolExecutionContext {
  agentId: string;
  taskId: string;
  toolSetIds: string[];
  /** Turn index for ordering metadata. */
  turnIndex?: number;
  /** Task-local event sequence. */
  eventSeq?: number;
  /** Allocate the next task-local sequence at lifecycle emission time. */
  nextEventSeq?: () => number;
  /** Parent assistant message ID for tool ordering. */
  parentMessageId?: string;
  /** Content block index in parent message. */
  contentIndex?: number;
  /** Dense tool call index among tool calls in the message. */
  toolCallIndex?: number;
  /** Stable internal identity, distinct from the provider's opaque call ID. */
  toolEntityId?: string;
}

/** Structured tool execution result. */
export interface ToolExecResult {
  ok: boolean;
  value?: unknown;
  error?: { code: string; message: string; retryable?: boolean };
}

/**
 * A ToolProvider is the discovery and execution adapter for one source of tools.
 * `ToolRegistryImpl.executeTool()` owns coordination around the provider: approval, lifecycle events,
 * timeout, cancellation, and structured results.
 */
export interface ToolProvider {
  id: string;
  source: ToolProviderSource;

  /** Discover available tools for the given context. */
  discover(context: ToolDiscoveryContext): Promise<ToolDef[]>;

  /** Execute a tool call. */
  execute(
    call: ToolCall,
    context: ToolExecutionContext,
    signal?: AbortSignal,
  ): Promise<ToolExecResult>;
}

File: packages/protocol/src/tools.rs
// ---- Protocol: tools — tool definitions, tool sets, policies ----

use serde::{Deserialize, Serialize};

// ---- Enums / string types ----

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ToolCapability {
    #[serde(rename = "workspace_read")]
    WorkspaceRead,
    #[serde(rename = "workspace_write")]
    WorkspaceWrite,
    #[serde(rename = "process")]
    Process,
    Network,
    #[serde(rename = "view_image")]
    ViewImage,
    #[serde(rename = "update_plan")]
    UpdatePlan,
    #[serde(rename = "user_input")]
    UserInput,
    Delegation,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ToolExposure {
    Direct,
    Deferred,
    Hidden,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ToolApprovalRequirement {
    Never,
    #[serde(rename = "on_request")]
    OnRequest,
    Always,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ToolProviderSource {
    #[serde(rename = "orch")]
    Orch,
    Host,
    #[serde(rename = "workspace")]
    Workspace,
    #[serde(rename = "mcp")]
    Mcp,
    Plugin,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ToolExecutionMode {
    Parallel,
    Sequential,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ToolSensitivity {
    Safe,
    Sensitive,
    Dangerous,
    Dynamic,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ToolApprovalPolicy {
    Never,
    #[serde(rename = "on_sensitive")]
    OnSensitive,
    Always,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ToolFailureMode {
    #[serde(rename = "return_error")]
    ReturnError,
    #[serde(rename = "fail_task")]
    FailTask,
}

// ---- ToolDef ----

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ToolExecutorRef {
    pub kind: String, // native, host, remote, sandbox, mcp, orchestrator
    pub target: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extra: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ToolMetadata {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "readOnly")]
    pub read_only: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub destructive: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "mutatesWorkspace")]
    pub mutates_workspace: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "producesArtifact")]
    pub produces_artifact: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ToolDef {
    pub name: String,
    pub description: String,
    #[serde(rename = "inputSchema")]
    pub input_schema: serde_json::Value,
    pub executor: ToolExecutorRef,
    #[serde(skip_serializing_if = "Option::is_none", rename = "executionMode")]
    pub execution_mode: Option<ToolExecutionMode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub exposure: Option<ToolExposure>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub capabilities: Option<Vec<ToolCapability>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approval: Option<ToolApprovalRequirement>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<ToolMetadata>,
}

// ---- ToolSets ----

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ToolSetMetadata {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source: Option<String>, // builtin, host, mcp, plugin, dynamic, agent
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
#[serde(rename_all = "camelCase")]
pub struct ToolPolicy {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sensitivity: Option<ToolSensitivity>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approval: Option<ToolApprovalPolicy>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "timeoutMs")]
    pub timeout_ms: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "executionMode")]
    pub execution_mode: Option<ToolExecutionMode>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "failureMode")]
    pub failure_mode: Option<ToolFailureMode>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ToolSetPolicy {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub defaults: Option<ToolPolicy>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "allowParallel")]
    pub allow_parallel: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "maxConcurrentCalls")]
    pub max_concurrent_calls: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum ToolSetToolRef {
    #[serde(rename = "provider_tool")]
    ProviderTool {
        #[serde(rename = "providerId")]
        provider_id: String,
        #[serde(rename = "toolName")]
        tool_name: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        alias: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        policy: Option<ToolPolicy>,
    },
    #[serde(rename = "provider_namespace")]
    ProviderNamespace {
        #[serde(rename = "providerId")]
        provider_id: String,
        namespace: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        alias: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        policy: Option<ToolPolicy>,
    },
    #[serde(rename = "orchestrator_control")]
    OrchestratorControl {
        action: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        alias: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        policy: Option<ToolPolicy>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ToolSet {
    pub id: String,
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub tools: Vec<ToolSetToolRef>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub policy: Option<ToolSetPolicy>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<ToolSetMetadata>,
}
